### DISCUSSION TITLE
Best way to read a bulk in endpoint

### DISCUSSION THREAD

**Discussion by brianreinhold** [pyusb/pyusb#430] [Q&A] [answered]
I note that there are two options for the bulk-in Endpoint..read() method; one where you specify the number of bytes to read and you get a buffer with the bytes read, and one where you provide a buffer and you get that buffer 'filled' and the number of bytes read returned.

I noted that there is a util method util.create_buffer(). This buffer is optimized to minimize the recreation of buffers every time. I do not fully understand that optimization. However, what I really want to know is whether this optimized buffer is something I should use in the read() method or is it already used in the read operation internally if I do not specify a buffer in the call?

So what is the most efficient approach to using the read method?

**Comment by jonasmalacofilho** [comment, marked as answer]
The purpose of `util.create_buffer()` is to provide a proper way to create a buffer for `read()` of the correct type.

To benefit from avoiding unnecessary allocations you call `util.create_buffer()` and manage the lifetime of the returned buffer yourself, and use that same buffer instance for many different `read()` operations.

**Reply by brianreinhold** [reply]
Thank you. New to this library (and Python) and so far, that is what I have done.
